### ISSUE TITLE
TypeError when connecting via mopidy-iris

### ISSUE THREAD

**Issue by KucharczykL** [jellyfin/mopidy-jellyfin#76]
I open the Iris web interface, navigate to Browse, click Jellyfin and there are only 3 items from my library. When I click on an item, the error below is displayed. 

Jellyfin is running behind a reverse proxy but in the config I define it by the server's IP so it shouldn't matter.

Jellyfin: 10.6.2 (Docker)
Mopidy             3.0.2
Mopidy-Iris        3.50.0
Mopidy-Jellyfin    0.9.1

The error:
```
ERROR    2020-08-04 19:00:53,364 [26685:Core-10] mopidy.core.library
  JellyfinBackend backend caused an exception.
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 17, in _backend_error_handling
    yield
  File "/usr/lib/python3.8/site-packages/mopidy/core/library.py", line 217, in lookup
    result = future.get()
  File "/usr/lib/python3.8/site-packages/pykka/_threading.py", line 45, in get
    _compat.reraise(*self._data['exc_info'])
  File "/usr/lib/python3.8/site-packages/pykka/_compat/__init__.py", line 29, in reraise
    raise value
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 193, in _actor_loop
    response = self._handle_receive(envelope.message)
  File "/usr/lib/python3.8/site-packages/pykka/_actor.py", line 299, in _handle_receive
    return callee(*message.args, **message.kwargs)
  File "/usr/lib/python3.8/site-packages/mopidy_jellyfin/library.py", line 45, in lookup
    contents = [
  File "/usr/lib/python3.8/site-packages/mopidy_jellyfin/library.py", line 46, in <listcomp>
    self.backend.remote.create_track(track)
  File "/usr/lib/python3.8/site-packages/mopidy_jellyfin/remote.py", line 473, in create_track
    artists=self.create_artists(track),
  File "/usr/lib/python3.8/site-packages/mopidy_jellyfin/remote.py", line 499, in create_artists
    return [
TypeError: 'NoneType' object is not iterable
```
My config file:
```
[jellyfin]
hostname = http://192.168.0.106:8096
username = lukas
password = snip
libraries = hudba
albumartistsort = True
```

Let me know if you need anything else.

**Comment by mcarlton00** [comment]
I see the problem.  I was only testing on music libraries when making the Iris fixes, but Audiobook libraries are still not behaving.

**Comment by KucharczykL** [comment]
Mine is a music library, though? I don't have any audio books.

**Comment by mcarlton00** [comment]
You're positive it's set as a Music library in the server and not accidentally not on book?

In that case, it's likely an issue with file structure and some poor assumptions in my code.  I might need to rework some code to decide behavior based more on item type instead of assuming structure.

**Comment by KucharczykL** [comment]
Just double-checked and yes, it is set as music library.

**Comment by mcarlton00** [comment]
I think I found it.  Turns out Iris has some interesting behavior under the hood.  Even when using the browse functionality, it's trying to do library lookups and falls back to the browse function.  In this case, it seems to have been failing on the lookup and never making it to the browse function.  These are generally different behaviors and accept different return types, so it's a bit confusing to be mixing them together like this.

**Comment by mcarlton00** [comment]
The console still spits out errors after my above fix, but it's only during tracklist events, they're coming from Iris itself, and everything appears to be working fine.

**Comment by smckend** [comment]
I seem to also be getting this issue with the latest 1.0.0 version. I'm using my own custom UI here which is working perfectly fine with spotify and soundcloud plugins.

```
Nov 15 18:34:03 raspberrypi mopidy[13271]: DEBUG    [JellyfinBackend-2] urllib3.connectionpool http://192.168.1.25:8096 "GET /Search/Hints?format=json&SearchTerm=sunday&IncludeItemTypes=Audio%2CMusicAlbum%2CMusicArtist HTTP/1.1" 200 618
Nov 15 18:34:03 raspberrypi mopidy[13271]: DEBUG    [JellyfinBackend-2] mopidy_jellyfin.http {'SearchHints': [{'ItemId': 'f864f149053c881763dbaece2a03f468', 'Id': 'f864f149053c881763dbaece2a03f468', 'Name': 'Sunday School', 'IndexNumber': 4, 'ProductionYear': 2018, 'ParentIndexNumber': 1, 'PrimaryImageTag': '1c1cb41052d9d203f460a4dc25d69676', 'Type': 'Audio', 'RunTimeTicks': 3177270016, 'MediaType': 'Audio', 'Album': 'Into The Morning', 'AlbumId': '42f427a827340263b84f3aad4778c8f3', 'AlbumArtist': 'Kyle Watson', 'Artists': ['Kyle Watson'], 'PrimaryImageAspectRatio': 1}, {'ItemId': '2e54112017b6bee3eec7a4f43c63c24b', 'Id': '2e54112017b6bee3eec7a4f43c63c24b', 'Name': 'Sunday Gentlemen', 'ProductionYear': 2013, 'Type': 'MusicAlbum', 'IsFolder': True, 'RunTimeTicks': 35910791168, 'AlbumArtist': 'Spit Syndicate', 'Artists': ['Drapht, Spit Syndicate', 'Illy, Spit Syndicate', 'Kai, Spit Syndicate', 'Sarah Corry, Spit Syndicate', 'Solo, Spit Syndicate', 'Spit Syndicate']}, {'ItemId': '799e12ce395a8cd72acdb3e2f91e8355', 'Id': '799e12ce395a8cd72acdb3e2f91e8355', 'Name': 'No Dope On Sundays (feat. Pusha T)', 'IndexNumber': 2, 'ProductionYear': 2017, 'ParentIndexNumber': 1, 'PrimaryImageTag': '063b688b6ccd93f52f31cfced7256941', 'Type': 'Audio', 'RunTimeTicks': 3830070016, 'MediaType': 'Audio', 'Album': 'No Dope On Sundays', 'AlbumId': 'b516880f866d00ffc306f7d98e2a0d9b', 'AlbumArtist': 'CyHi The Prynce', 'Artists': ['CyHi The Prynce'], 'PrimaryImageAspectRatio': 1}, {'ItemId': 'b516880f866d00ffc306f7d98e2a0d9b', 'Id': 'b516880f866d00ffc306f7d98e2a0d9b', 'Name': 'No Dope On Sundays', 'ProductionYear': 2017, 'PrimaryImageTag': '7092964f2c0e423c655bd85fe715faea', 'Type': 'MusicAlbum', 'IsFolder': True, 'RunTimeTicks': 43926990848, 'AlbumArtist': 'Cyhi the Prynce', 'Artists': ['CyHi The Prynce'], 'PrimaryImageAspectRatio': 1}], 'TotalRecordCount': 4}
Nov 15 18:34:03 raspberrypi mopidy[13271]: INFO     [JellyfinBackend-2] pykka Exception returned from JellyfinBackend (urn:uuid:35255bf5-4bfd-46d1-a242-500dc6c9e09b) to caller:
Nov 15 18:34:03 raspberrypi mopidy[13271]: Traceback (most recent call last):
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/utils.py", line 24, in _memoized
Nov 15 18:34:03 raspberrypi mopidy[13271]:     value, last_update = self.cache[args]
Nov 15 18:34:03 raspberrypi mopidy[13271]: TypeError: unhashable type: 'dict'
Nov 15 18:34:03 raspberrypi mopidy[13271]: During handling of the above exception, another exception occurred:
Nov 15 18:34:03 raspberrypi mopidy[13271]: Traceback (most recent call last):
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/lib/python3/dist-packages/pykka/_actor.py", line 193, in _actor_loop
Nov 15 18:34:03 raspberrypi mopidy[13271]:     response = self._handle_receive(envelope.message)
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/lib/python3/dist-packages/pykka/_actor.py", line 299, in _handle_receive
Nov 15 18:34:03 raspberrypi mopidy[13271]:     return callee(*message.args, **message.kwargs)
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/library.py", line 86, in search
Nov 15 18:34:03 raspberrypi mopidy[13271]:     return self.backend.remote.search(query)
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/utils.py", line 39, in _memoized
Nov 15 18:34:03 raspberrypi mopidy[13271]:     return self.func(*args)
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/remote.py", line 684, in search
Nov 15 18:34:03 raspberrypi mopidy[13271]:     tracks.append(self.create_track(item))
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/remote.py", line 573, in create_track
Nov 15 18:34:03 raspberrypi mopidy[13271]:     artists=self.create_artists(track=track),
Nov 15 18:34:03 raspberrypi mopidy[13271]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/remote.py", line 606, in create_artists
Nov 15 18:34:03 raspberrypi mopidy[13271]:     for artist in track.get('ArtistItems')
Nov 15 18:34:03 raspberrypi mopidy[13271]: TypeError: 'NoneType' object is not iterable
Nov 15 18:34:03 raspberrypi mopidy[13271]: WARNING  [Core-10] mopidy.core.library JellyfinBackend does not implement library.search() with "exact" support. Please upgrade it.
```

**Comment by mcarlton00** [comment]
Hmm, I can't replicate it in my library.  It seems to be a matter of bad/missing metadata.  The only way I can see for it to fail at that particular point is if a given item doesn't have an `ArtistItems` field, which doesn't make any sense.  Though I also don't see a reason to be pulling `ArtistItems` instead of just `Artists`.  If you're up for a little debugging, I believe we can confirm that and find a fix pretty quickly.

Can you change lines 601-607 of `/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/remote.py` to match this and see if that fixes the issue? (the only changes are the ends of lines 604 and 606)

```python
601         if track:
602             return [
603                 models.Artist(
604                     name=name if name else artist
605                 )
606                 for artist in track.get('Artists', [])
607             ]
```

**Comment by smckend** [comment]
Thanks for the quick response. Always up for some debugging.

The change suggested helped for the query that produced the earlier issue but upon adding more than 1 query item it throws the following:

```
Nov 15 20:05:13 raspberrypi mopidy[22800]: INFO     [JellyfinBackend-2] pykka Exception returned from JellyfinBackend (urn:uuid:99c9be47-bd4e-41bb-b2ce-23f86bd0b6b6) to caller:
Nov 15 20:05:13 raspberrypi mopidy[22800]: Traceback (most recent call last):
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/utils.py", line 24, in _memoized
Nov 15 20:05:13 raspberrypi mopidy[22800]:     value, last_update = self.cache[args]
Nov 15 20:05:13 raspberrypi mopidy[22800]: TypeError: unhashable type: 'dict'
Nov 15 20:05:13 raspberrypi mopidy[22800]: During handling of the above exception, another exception occurred:
Nov 15 20:05:13 raspberrypi mopidy[22800]: Traceback (most recent call last):
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/lib/python3/dist-packages/pykka/_actor.py", line 193, in _actor_loop
Nov 15 20:05:13 raspberrypi mopidy[22800]:     response = self._handle_receive(envelope.message)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/lib/python3/dist-packages/pykka/_actor.py", line 299, in _handle_receive
Nov 15 20:05:13 raspberrypi mopidy[22800]:     return callee(*message.args, **message.kwargs)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/library.py", line 86, in search
Nov 15 20:05:13 raspberrypi mopidy[22800]:     return self.backend.remote.search(query)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/utils.py", line 39, in _memoized
Nov 15 20:05:13 raspberrypi mopidy[22800]:     return self.func(*args)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/local/lib/python3.7/dist-packages/mopidy_jellyfin/remote.py", line 699, in search
Nov 15 20:05:13 raspberrypi mopidy[22800]:     albums=albums
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/lib/python3/dist-packages/mopidy/models/immutable.py", line 159, in __call__
Nov 15 20:05:13 raspberrypi mopidy[22800]:     instance = super().__call__(*args, **kwargs)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/lib/python3/dist-packages/mopidy/models/immutable.py", line 35, in __init__
Nov 15 20:05:13 raspberrypi mopidy[22800]:     self._set_field(key, value)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/lib/python3/dist-packages/mopidy/models/immutable.py", line 188, in _set_field
Nov 15 20:05:13 raspberrypi mopidy[22800]:     object.__setattr__(self, name, value)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/lib/python3/dist-packages/mopidy/models/fields.py", line 50, in __set__
Nov 15 20:05:13 raspberrypi mopidy[22800]:     value = self.validate(value)
Nov 15 20:05:13 raspberrypi mopidy[22800]:   File "/usr/lib/python3/dist-packages/mopidy/models/fields.py", line 176, in validate
Nov 15 20:05:13 raspberrypi mopidy[22800]:     f"Expected {self._name} to be a collection of "
Nov 15 20:05:13 raspberrypi mopidy[22800]: TypeError: Expected artists to be a collection of Artist, not [[], []]
```

**Comment by mcarlton00** [comment]
I haven't replicated that, but I have found a different error that I'm seeing when using Iris involving the search functions.  How are you calling the search?  They might be related

**Comment by smckend** [comment]
Like this: `{"jsonrpc":"2.0","id":"ef3ffeeb-45f6-41ab-8cd9-4bbd65d88f58","method":"core.library.search","params":{"query":{"any":["some","query"]},"uris":[],"exact":false}}`

**Comment by mcarlton00** [comment]
Alright I see a problem.  Seems likely to be the root cause for both issues.  Working on rewriting that search function now, so hopefully we'll have a fix shortly

**Comment by smckend** [comment]
Thanks for your help

**Comment by mcarlton00** [comment]
I've just released version 1.0.1 that should fix this
